### PR TITLE
Improve and finalize availability

### PR DESCRIPTION
**Pull request by Hlavtox**
| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | 9.0.x
| Description?      | I finished the remaining things regarding availability. See below.
| Type?             | improvement
| Category?         | CO
| BC breaks?        | yes
| Deprecations?     | no
| How to test?      | See below
| UI Tests          | 
| Fixed issue or discussion?     | 
| Related PRs       | https://github.com/PrestaShop/hummingbird/pull/658, https://github.com/PrestaShop/classic-theme/pull/159
| Sponsor company   | 

### 1. Label when out of stock behavior
Product `Label when out of stock` that is configurable for each product, was only used when backorders were enabled for the product. If they were disabled, it only looked on the value in `Shop settings > Product settings`. This is insufficient, because we can't specify the availability for these products.

Example - `We are expecting the delivery in 10/2024.` or `There are problems on manufacturer side, availability on request.` and many other variants.

So, I slightly tweaked it and now this field is used for both cases. If it's empty, it will look to the prestashop settings value.

Before, I was thinking into adding another field to a product, but that's not needed. One field is sufficient. :-)

### 2. Helper labels in admin
Both `Label when in stock` and `Label when out of stock` now have helper labels under them with a link to shop settings, explaining what they do and when they are displayed. Both in combination and product forms.

#### Combination
- `Label when in stock` - `This will be the displayed availability of the combination, if there is at least 1 in stock. If you don\'t enter anything, the base value set on the product will be used.` 
- `Label when out of stock` - `This will be the displayed availability of the combination, if it\'s not in stock. If you don\'t enter anything, the base value set on the product will be used.`

#### Product
- `Label when in stock` - `This will be the displayed availability of the product, if there is at least 1 in stock. If you don\'t enter anything, value from [1]Shop Parameters > Product Settings[/1] will be used.`
- `Label when out of stock` - `This will be the displayed availability of the product, if it\'s not in stock. If you don\'t enter anything, value from [1]Shop Parameters > Product Settings[/1] will be used.`

#### Shop Parameters > Product settings
- `Label of in-stock products` - `This will be the default displayed availability of a product, if there is at least 1 in stock. If you don\'t enter anything, nothing will be displayed. Further customization is possible for each product.`
- `Label of out-of-stock products with allowed backorders` - `This will be the default displayed availability of a product, if it\'s not in stock and backordering it is enabled. If you don\'t enter anything, nothing will be displayed. Further customization is possible for each product.`
- `Label of out-of-stock products with denied backorders` - `This will be the default displayed availability of a product, if it\'s not in stock and backordering it is denied. If you don\'t enter anything, nothing will be displayed. Further customization is possible for each product.`

### 3. availability_submessage
I moved `Product available with different options` into a submessage key, as we originally designed it for hummingbird. This will be a new smaller text under the availability.

See specs - https://github.com/PrestaShop/prestashop-specs/blob/master/content/1.7/back-office/shop-parameters/product-settings/page.md - 

> if at least one other combination than the default one (on the product listings) or the selected one (on product page) can be ordered, the "Other product variations available" message will be displayed under the "Label of out-of-stock products with denied backorders" message.

### 4. How to test
- Go to `Shop settings > Product settings` and set some default texts to both `Label of out-of-stock products with allowed backorders` and `Label of out-of-stock products with denied backorders`.
- Create product with quantity 0, set some text to `Label when out of stock` field on the product.
- Enable backorders, go to frontoffice and see that you see the `Label when out of stock` of the product.
- Disable backorders, go to frontoffice and see that you see the `Label when out of stock` of the product.
- Remove text from `Label when out of stock` on the product.
- Enable backorders, go to frontoffice and see that you see the default text from Shop settings > Product settings.
- Disable backorders, go to frontoffice and see that you see the default text from Shop settings > Product settings.

### CODE REVIEW

**Comment by ps-jarvis** [comment]
This pull request seems to contain new translation strings. I have summarized them below to ease up review:

- `Admin.Catalog.Feature`
    - [ ] `Label when out of stock`

(Note: this is an automated message, but answering it will reach a real human)

**Review by jolelievre — DISMISSED** [review]
Thanks @Hlavtox

**Review by ibahloul-ps — DISMISSED** [review]
All good to me!
Thanks for your contribution @Hlavtox :)

**Review by tleon — DISMISSED** [review]

**Review by boherm — DISMISSED** [review]

**Comment by florine2623** [comment]
I've relaunched the automated tests multiple times, they are always failing : https://github.com/florine2623/testing_pr/actions/runs/11126152612

Could you check @Hlavtox and @PrestaShop/qa-automation ?

Thanks!

**Comment by Progi1984** [comment]
@florine2623 If you have relaunched multiples times, it's not UI tests which have problem, it's code which have regressions :wink:

**Comment by Hlavtox** [comment]
@florine2623 

New run - https://github.com/Hlavtox/ga.tests.ui.pr/actions/runs/11141871170

**Comment by Hlavtox** [comment]
@Progi1984 Are tests really working? This is not caused by my PR:

>   1) BO - Shop Parameters - Search : Maximum word length (in characters)
>        should set the maximum word length to "":
>      AssertionError: expected 'Please fill out this field.' to include 'Please fill in this field.'

**Comment by Hlavtox** [comment]
@PrestaShop/qa-automation Hi guys, I tried to fix the UI tests, but it's trial and error. Could you please adapt them to this PR? :-)

Basically, if a product is not in stock and backorders are disabled:
- Before: Label in PS settings
- Now: Label on combination > Label on product > Label in PS settings

**Review by kpodemski — DISMISSED** [review]
Hello

We recently introduced the 9.0.x branch, which will be the base for PrestaShop 9.0. If you want to see your change in this version, please rebase your PR onto 9.0.x. The `develop` will be used for PrestaShop 9.1, so you need to decide what to do.

Please rebase onto `9.0.x` or let me know your decision if you don't need to have it there so I can unblock the pull request.

**Comment by Hlavtox** [comment]
@kpodemski It's ready.

@PrestaShop/qa-automation Guys can you please adapt the UI tests? :-)

**Comment by Progi1984** [comment]
@Hlavtox May be in next weeks.

**Comment by Hlavtox** [comment]
@Progi1984 Is there a chance to do it sooner than "weeks" so we can merge it into 9.0.x?

It's relatively simple PR. :-)

Maybe @nesrineabdmouleh could have time? 🙏

**Comment by Hlavtox** [comment]
https://github.com/Hlavtox/ga.tests.ui.pr/actions/runs/11540232276

**Comment by Hlavtox** [comment]
@kpodemski Krystian, this PR is currently stuck for a month already due to tests failing that needs adapting. Can we merge it? I can't get hold of anyone that would be responsible for these tests, unfortunately.

**Comment by nesrineabdmouleh** [comment]
Hello @kpodemski @Hlavtox,

Sorry for the delay, this is a new run https://github.com/nesrineabdmouleh/testing_pr/actions/runs/11572718762
I will adapt tests if there are failures. :)

Thanks!

**Comment by Hlavtox** [comment]
@nesrineabdmouleh Thank you! :-)

Needs two fixes:
- More precise selector in one test in BO, because I added a new link and now it doesn't know which one to click.
- In FO, there is now a different stock text - expected ' Out of stock' to include 'Out-of-Stock'

**Comment by nesrineabdmouleh** [comment]
New run https://github.com/nesrineabdmouleh/testing_pr/actions/runs/11592778818

**Review by ShaiMagal — DISMISSED** [review]

**Comment by Hlavtox** [comment]
@nesrineabdmouleh @Progi1984 And one approval here please :-) The run is all green. I rebased it.

**Review by nesrineabdmouleh — DISMISSED** [review]

File: src/Adapter/Presenter/Product/ProductLazyArray.php
Comment (kpodemski) [inline src/Adapter/Presenter/Product/ProductLazyArray.php:1164]: so, hummingbird PR to follow, yes?

**Review by kpodemski — DISMISSED** [review]

**Comment by ps-jarvis** [comment]
Hi, thanks for this contribution!

I found some issues with the Pull Request description:

* The `branch` should be `develop` or `8.2.x`. ([Read explanation](https://devdocs.prestashop-project.org/9/contribute/contribution-guidelines/pull-requests/#branch))

Would you mind having a look at it? This will help us understand how interesting your contribution is, thank you very much!

(Note: this is an automated message, but answering it will reach a real human)

File: src/Adapter/Presenter/Product/ProductLazyArray.php
Comment (Hlavtox) [reply src/Adapter/Presenter/Product/ProductLazyArray.php:1164]: https://github.com/PrestaShop/hummingbird/pull/658

**Review by Progi1984 — DISMISSED** [review]

**Comment by Hlavtox** [comment]
Ping @florine2623 , this can now be tested. :-)

Ideally, together with https://github.com/PrestaShop/hummingbird/pull/658 so you see the change in front office.

**Comment by Hlavtox** [comment]
Ping @PrestaShop/qa-functional for testing. :-) It's a small PR basically only changing one stock message.

**Review by florine2623 — DISMISSED** [review]
Hello @Hlavtox ,

I have tested this PR with https://github.com/PrestaShop/hummingbird/pull/658.

I had an issue with installing the PR, so I followed your suggestion to use a build of hummingbird 0.2.0 and modify just that one file.

I followed the How to test steps.
- Enable backorders, go to frontoffice and see that you see the `Label when out of stock` of the product.
Screenshot 2024-12-02 at 15 37 24

- Disable backorders, go to frontoffice and see that you see the `Label when out of stock` of the product.
Screenshot 2024-12-02 at 15 36 51

- Enable backorders, go to frontoffice and see that you see the default text from Shop settings > Product settings.
Screenshot 2024-12-02 at 15 32 08

- Disable backorders, go to frontoffice and see that you see the default text from Shop settings > Product settings.
Screenshot 2024-12-02 at 15 32 33

- In stock 

Screenshot 2024-12-02 at 15 40 34

Works as expected !
Thanks @Hlavtox 💪

**Comment by Hlavtox** [comment]
Thank you @florine2623! 🌹🌷I will just restart the tests to let the new ones pass. :-)

**Comment by florine2623** [comment]
@Hlavtox ^^

Before merging, I have tested with classic, I have an issue actually 😭 

I have tested with a product with stock = 0
I have added a label in Shop parameters > Product Settings for `out-of-stock products with denied orders`, the label is displayed as a tag on the product image.

I should not have a tag, but just a label under the quantity. Do you agree ?


Screenshot 2024-12-02 at 15 57 13

Screenshot 2024-12-02 at 15 55 46

In my case here, I can have 2 different message for the same product/product stock. It will be confusing for the customer.

Second issue, when I have some stock available, the label is also confusing. Because it is a 🚫  sign but I can still buy the product.

Screenshot 2024-12-02 at 16 00 28


What do you think ?

**Comment by Hlavtox** [comment]
@florine2623 

> I have added a label in Shop parameters > Product Settings for out-of-stock products with denied orders, the label is displayed as a tag on the product image.

Issue with out of stock label is fixed. :-) I didn't even know it was dynamic. Sorry!

Snímek obrazovky 2024-12-02 161302
Snímek obrazovky 2024-12-02 161324

> Second issue, when I have some stock available, the label is also confusing. Because it is a 🚫 sign but I can still buy the product.

Will be fixed by another PR for classic, like I did for hummingbird. I will ping you later for testing. :-)

**Review by ShaiMagal — APPROVED** [review]

**Comment by Hlavtox** [comment]
@florine2623 Ready to retest. I also made a classic theme PR that makes it look nice. :-)

https://github.com/PrestaShop/classic-theme/pull/159

So you can test it there and see that it looks as good as on hummingbird.

**Review by kpodemski — APPROVED** [review]

**Review by florine2623 — APPROVED** [review]
Hello @Hlavtox ,

Thanks for the PR !

I tested with classic theme https://github.com/PrestaShop/classic-theme/pull/159.

It looks as expected :)
I tested with stock label directly in product page as well as in product settings.
I checked the tag added to the image product when stock is modified ✅ 

When in stock - in Product page :
Screenshot 2024-12-12 at 12 02 13

When in stock - in Product settings page :
Screenshot 2024-12-12 at 12 02 50

When out of stock, denied orders - in Product page :
Screenshot 2024-12-12 at 12 03 46

When out of stock, allowed orders - in Product page :
Screenshot 2024-12-12 at 12 12 19


When out of stock, denied orders - in Product settings page :
Screenshot 2024-12-12 at 12 07 04

When out of stock, allowed orders - in Product settings page :
Screenshot 2024-12-12 at 12 07 28


LGTM ! Thanks :)

**Review by Progi1984 — APPROVED** [review]

**Pull request merged**
